File: 0240-search-a-2d-matrix-ii/0240-search-a-2d-matrix-ii.py
class Solution:
    def searchMatrix(self, g: List[List[int]], x: int) -> bool:
        if not g : return False
        r = len(g)
        c = len(g[0])

        def recs(l,u,r,d):
            if l>r or u>d:
                return False
            elif x<g[u][l] or x>g[d][r]:
                return False
            mid = (r-l)//2 + l

            row = u
            while row<=d and g[row][mid] <= x:
                if g[row][mid] ==x: return True
                row+=1
            
            return recs(l,row,mid-1,d) or recs(mid+1,u,r,row-1)

        return recs(0,0,c-1,r-1)



        

File: 0240-search-a-2d-matrix-ii/0240-search-a-2d-matrix-ii.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& g, int x) {
        
        int r = g.size()-1;
        int c = 0;

        while(true){
            int lastele = g[r][c];
            if(lastele==x)return true;
            if(lastele>x){r--;}
            if(lastele<x)c++;
            if(r<0 || c>=g[0].size())break;
        }

        return false;
    }
};